contracts/test/PlonkVerifier.t.sol

```diff
@@ -0,0 +1,432 @@
+        cmds[2] = vm.toString(uint32(1));
+
+        bytes memory result = vm.ffi(cmds);
+        (
+            IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+            uint256[][] memory publicInputs,
+            IPlonkVerifier.PlonkProof[] memory proofs,
+            bytes[] memory extraTranscriptInitMsgs
+        ) = abi.decode(
+            result,
+            (IPlonkVerifier.VerifyingKey[], uint256[][], IPlonkVerifier.PlonkProof[], bytes[])
+        );
+
+        extraTranscriptInitMsgs[0] = badMsg;
+        assert(!V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs));
+    }
+}
+
+contract PlonkVerifier_validateProof_Test is PlonkVerifierCommonTest {
+    /// @dev Test that a valid proof shouldn't revert
+    function test_validProof_succeeds() external {
+        // a valid proof
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(42);
+
+        V._validateProof(proof);
+    }
+
+    /// forge-config: default.fuzz.runs = 30
+    /// @dev Randomly pick a coordinate of a point among points in a proof
+    /// mutate it to another value so that the point is no longer valid,
+    /// test if our check will revert.
+    function testFuzz_RevertIfProofContainsInvalidGroup(uint256 nthPoint, bool testX) external {
+        // a valid proof
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(42);
+
+        // we are testing the `nthPoint` in the `proof`,
+        // only mutating a single field element (either x or y coordinate)
+        // There are 13 points in total.
+        nthPoint = bound(nthPoint, 0, 12);
+
+        assembly {
+            if testX {
+                // muteate the x coordinate
+                mstore(mload(add(proof, mul(0x20, nthPoint))), 0x1234)
+            }
+            // else, mutate y coordinate
+            mstore(add(mload(add(proof, mul(0x20, nthPoint))), 0x20), 0x1234)
+        }
+
+        vm.expectRevert();
+        V._validateProof(proof);
+    }
+
+    /// forge-config: default.fuzz.runs = 15
+    /// @dev Randomly pick field in a proof mutate it to invalid value
+    /// test if our check will revert.
+    function testFuzz_RevertIfProofContainsInvalidField(uint256 nthField) external {
+        // a valid proof
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(42);
+        uint256 invalidField = BN254.R_MOD;
+
+        // we are testing the `nthField` in the `proof`,
+        // There are 10 points in total (with 13 points in front)
+        nthField = bound(nthField, 0, 9);
+
+        assembly {
+            let start := add(proof, mul(0x20, 13))
+            mstore(add(start, mul(nthField, 0x20)), invalidField)
+        }
+
+        vm.expectRevert();
+        V._validateProof(proof);
+    }
+}
+
+contract PlonkVerifier_preparePcsInfo_Test is PlonkVerifierCommonTest {
+    /// forge-config: default.fuzz.runs = 5
+    /// @dev Test `preparePcsInfo` matches that of Jellyfish
+    function testFuzz_preparePcsInfo_matches(
```

Comment:
```
Encountered 1 failing test in contracts/test/PlonkVerifier.t.sol:PlonkVerifier_preparePcsInfo_Test
[FAIL. Reason: InvalidPolyEvalArgs() Counterexample: calldata=0x0x99fdc07100000000000000000000000000000000000000000000000000000000000005fe000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000003b2, args=[1534, [], 0x00000000000000000000000000000000000000000000000000000000000003b2]] testFuzz_preparePcsInfo_matches(uint64,uint256[],bytes) (runs: 1, μ: 8001859, ~: 8001859)
```

Reply:
good catch, I will add test logic for empty `publicInput` array.

Reply:
done in [90bf8d1](https://github.com/EspressoSystems/espresso-sequencer/pull/722/commits/90bf8d1c42a6613213ef9efe4711a57e3c803a89)